File: binary_trees/cpp/binarytree.cpp
#include <iostream>
#include <memory>

class BinaryTree {
private:
  std::unique_ptr<BinaryTree> left;
  std::unique_ptr<BinaryTree> right;

public:
  BinaryTree(std::unique_ptr<BinaryTree> left = nullptr,
             std::unique_ptr<BinaryTree> right = nullptr)
      : left(std::move(left)), right(std::move(right)) {}

  ~BinaryTree() {
    left = nullptr;
    right = nullptr;
  }

  long countNodes() const {
    if (left == nullptr) {
      return 1;
    } else {
      return 1 + left->countNodes() + right->countNodes();
    }
  }

  static std::unique_ptr<BinaryTree> createBottomUp(unsigned int depth) {
    if (depth > 0) {
      return std::make_unique<BinaryTree>(createBottomUp(depth - 1),
                                          createBottomUp(depth - 1));
    } else {
      return std::make_unique<BinaryTree>();
    }
  }
};

int main(int argc, char *argv[]) {
  if (argc != 2) {
    std::cerr << "Usage: binarytree [depth]\n";
    return 1;
  }
  unsigned int N, minDepth, maxDepth;
  N = std::atol(argv[1]);

  minDepth = 4;
  if (minDepth + 2 > N) {
    maxDepth = minDepth + 2;
  } else {
    maxDepth = N;
  }

  unsigned int stretchDepth = maxDepth + 1;
  std::unique_ptr<BinaryTree> stretchTree = BinaryTree::createBottomUp(stretchDepth);
  std::cout << "stretch tree of depth " << stretchDepth
            << "\t check: " << stretchTree->countNodes() << std::endl;

  std::unique_ptr<BinaryTree> longLivedTree = BinaryTree::createBottomUp(maxDepth);

  long check = 0;
  long iterations;
  for (unsigned int depth = minDepth; depth <= maxDepth; depth += 2) {
    iterations = 1L << (maxDepth - depth + minDepth);
    check = 0;
    for (long i = 1; i <= iterations; i++) {
      std::unique_ptr<BinaryTree> tempTree = BinaryTree::createBottomUp(depth);
      check += tempTree->countNodes();
    }
    std::cout << iterations << "\t trees of depth " << depth
              << "\t check: " << check << std::endl;
  }
  std::cout << "long lived tree of depth " << maxDepth
            << "\t check: " << longLivedTree->countNodes() << std::endl;

  return 0;
}

File: binary_trees/ocaml/binarytree.ml
type binary_tree = Empty | Node of binary_tree * binary_tree

let rec count_nodes = function
  | Node (left, right) -> 1 + count_nodes left + count_nodes right
  | Empty -> 0

let rec create_bottom_up depth =
  if depth >= 0 then
    Node (create_bottom_up (depth - 1), create_bottom_up (depth - 1))
  else Empty

let () =
  if Array.length Sys.argv <> 2 then (
    Printf.eprintf "Usage: binarytree <depth>\n";
    exit 1);

  let n = int_of_string Sys.argv.(1) in
  let min_depth = 4 in
  let max_depth = max (min_depth + 2) n in

  let stretch_depth = max_depth + 1 in
  let stretch_tree = create_bottom_up stretch_depth in
  Printf.printf "stretch tree of depth %d\t check: %d\n" stretch_depth
    (count_nodes stretch_tree);

  let long_lived_tree = create_bottom_up max_depth in

  let rec loop_depth depth =
    if depth <= max_depth then (
      let iterations = 1 lsl (max_depth - depth + min_depth) in
      let check = ref 0 in
      for _ = 1 to iterations do
        let temp_tree = create_bottom_up depth in
        check := !check + count_nodes temp_tree
      done;
      Printf.printf "%d\t trees of depth %d\t check: %d\n" iterations depth
        !check;
      loop_depth (depth + 2))
  in
  loop_depth min_depth;

  Printf.printf "long lived tree of depth %d\t check: %d\n" max_depth
    (count_nodes long_lived_tree)
